### ISSUE TITLE
picture_viewer plugin doesn't restore saved scrolling positions properly

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#3170]
See PR #3171 for my (futile!) experiments.

I can see what is happening: scrolling a picture can somehow throw off the scrolling of the next slide (when moving forward) or the previous slide (when moving backward).

This is only an issue when `--use-db` is in effect. In that case the picture viewer uses an internal dict and saves that dict to a `.json` file on exit. I'm pretty sure the dict itself is updated correctly. The problem is in moving the viewrect (viewport?) just after creating the image.

@tbpassin I wonder whether you would consider helping with this issue. I've tried everything I can think of with Qt scrolling, but I'm a putz when it comes to graphics.

**Comment by tbpassin** [comment]
I could look at it, though I'm mostly in the dark about scrolling.  I do know that the QtWebEngine doesn't give you an API for scrollbar positioning, as best as I have been able to find out.  That's been  maddening in the VR3 context.  The older WebKit widget did.

Can you point me to some example cases so I don't have to roll my own?

**Comment by edreamleo** [comment]
@tbpassin I'm not sure I know what you mean by example cases. 

`Slides.scrollContentsBy` gets called when the viewRect has _already_ scrolled.  See the docs for [QAbstractScrollArea.scrollContentsBy](https://doc.qt.io/qt-6/qabstractscrollarea.html#scrollContentsBy) which says:

"Calling this function in order to scroll programmatically is an error, use the scroll bars instead (e.g. by calling [QScrollBar::setValue](https://doc.qt.io/qt-6/qabstractslider.html#value-prop)() directly)."

The main thing is that `Slides.do_scroll` isn't working :-)

**Comment by tbpassin** [comment]
After looking at the docstring I saw I just had to make a directory with some images.  I'm running the pictureviewer from a node in my workbook.

As for scrolling, I'm not sure at this point what the problem is.  If I zoom a picture so it's too large for the window, I can pan it using the arrow keys.  But I don't see any scrollbars.  How is panning different from scrolling?

I do have an unrelated problem, though.  Two of my three test pictures display rotated 90 deg.  They don't show rotated in other viewers I tried, including Irfanview, Paint, and the Windows picture viewer.  At a minimum I think we'll want to add a command to rotate the pictures.

**Comment by tbpassin** [comment]
Finally I think I'm getting synced up with you.  When I specify _use_db_ to `run()`, then close and re-open the viewer, the zoom factor was restored but the panned location was not.

Is this panned location what you want to fix?

**Comment by edreamleo** [comment]
@tbpassin Yes. I want to restore what you are calling the panned location.

**Comment by tbpassin** [comment]
It's definitely at the `scrollContentsBy()` - the right dx, dy args are there, but it does nothing.  Maybe we'll have to instantiate scroll bars and tell them to scroll.  More reading and experimentation to do ...

**Comment by edreamleo** [comment]
@tbpassin You're on the right track, but remember that the Qt documentation explicitly states that `scrollContentsBy` is *not* the proper place to do the actual scrolling. In fact, the default (base class) version of `scrollContentsBy` just does an `update`. The `dx` and `dy` args are for information only, and the overridden `Slides.scrollContentsBy` uses the args in the call to `Slides.save_data`.

In other words, `Slides.scrollContentsBy` *only* calls `Slides.save_data`. The solution, I think, must be in `Slides.do_scroll`.

Btw, it's conventional in Leo to use underscore name (do_scroll) to ensure that there is no conflict with Qt names, which are *always* camelCase. I don't think there is any Qt `doScroll` method or ivar, but it pays to be safe.

**Comment by tbpassin** [comment]
If we don't disable the scrollbars, then it works as expected (need to reverse the sign of self.dy):
```
# Scroll the pixmap.
if self.dx or self.dy:
    self.scroll_area.horizontalScrollBar().setValue(self.dx)
    self.scroll_area.verticalScrollBar().setValue(-self.dy)
```
We don't even need to write any new handlers - they come built in.

I prefer having the scroll bars anyway- they are what anyone would expect to have in an app like this.

**Comment by edreamleo** [comment]
@tbpassin Thanks Thomas. I'll try it out.

**Comment by edreamleo** [comment]
@tbpassin The problem has always seemed to be an interaction between the zoom value and the dx/dy offsets.  I think the problem is inherently tricky. I've even considered re-initing the widgets or the layout for each slide. That's also tricky :-)

**Comment by tbpassin** [comment]
I was working with the code in devel.   Here is my version of the relevant parts:

[show_slide.zip](https://github.com/leo-editor/leo-editor/files/10819059/show_slide.zip)

Also (not in this outline) I commented out the code that removed the scrollbars.

You didn't say anything about the zoom acting up, so I didn't look at that.  But with my change and --use-db, the zoom seems to be remembered and restore, at least to my casual glance.

**Comment by edreamleo** [comment]
Closed via PR #3171: This issue is not worth any more work right now.

**Comment by tbpassin** [comment]
If you went with my version of the code the pan/zoom would be remembered as desired (except for the rotation issue) - both during moving from slide to slide and after restart.

Perhaps there is another behavior I am not aware of?

**Comment by edreamleo** [comment]
@tbpassin We are probably seeing different behavior. I've stopped work on this issue for now.
